core/src/rest/http_client.ts

```diff
@@ -0,0 +1,240 @@
+    this.version = options?.version ?? DISCORD_API_VERSION;
+    this.maxRetries = options?.maxRetries ?? this.maxRetries;
+  }
+
+  async request(
+    method: HTTPMethod,
+    endpoint: Endpoint,
+    options?: RequestOptions,
+  ) {
+    const bucket = options?.bucket ?? getRouteBucket(endpoint);
+    const queue = this.queue.get(bucket) ??
+      (this.queue.set(bucket, new Queue()), this.queue.get(bucket)!);
+
+    const execute = async (): Promise<any> => {
+      await queue.wait();
+
+      // Enqueue it and wait for the queue to be resolved.
+      try {
+        const headers: Record<string, string> = {
+          "user-agent": this.userAgent,
+        };
+
+        if (this.token !== undefined && this.tokenType !== undefined) {
+          headers["authorization"] = `${this.tokenType} ${this.token}`.trim();
+        }
+
+        if (options?.headers !== undefined) {
+          Object.assign(headers, options.headers);
+        }
+
+        let body = undefined;
+        if (
+          ["POST", "PATCH", "PUT"].includes(method)
+        ) {
+          // It's multipart formdata.
+          if (options?.body instanceof FormData) {
+            body = options.body;
+          } else {
+            if (typeof options?.body === "object") {
+              // It's a simple JSON body.
+              body = JSON.stringify(
+                options.body,
+                (_, v) => typeof v === "bigint" ? v.toString() : v,
+              );
+              headers["content-type"] = "application/json";
+            } else {
+              // Default to empty body to avoid Cloudflare errors.
+              body = "";
+            }
+          }
+        }
+
+        const res = await fetch(
+          `${this.baseURL}/v${this.version}${endpoint}`,
+          {
+            method,
+            headers,
+            body,
+          },
+        );
+
+        const resetTime = Date.now() +
+          Number(res.headers.get("X-RateLimit-Reset-After")) * 1000 + 5; // add a little bit of delay to avoid 429s
+        queue.resetTime = resetTime;
+        if (res.headers.get("X-RateLimit-Global") === "true") {
+          this.queue.forEach((q) => {
+            if (q.resetTime < resetTime) {
+              q.resetTime = resetTime;
+            }
+          });
+        }
+
+        if (res.ok) {
+          return await res.json();
+        } else if (res.status >= 500 && res.status < 600) {
+          await res.body?.cancel();
+          throw new Error(
+            `Discord API Internal Server Error (${res.status})`,
+          );
+        } else if (res.status >= 400 && res.status < 500) {
+          await res.body?.cancel();
+          throw new Error(
+            `Discord API Error (${res.status})`,
+          );
+        } else {
+          await res.body?.cancel();
+          throw new Error(
+            `What in the world is this code? (${res.status})`,
+          );
+        }
+      } catch (e) {
```

Comment:
Suggestion: Maybe close the `try {} catch` before processing the request?
Anything either is okay or throws a error. So rethrowing wouldn't be necesarry here if `queue.shift()` also get's moved to before processing the response

Reply:
that's a good point actually